File: test/test_interfaces/unit/test_obc_gs_ax25.cpp
#include "obc_gs_ax25.h"
#include "obc_gs_fec.h"
#include "obc_gs_errors.h"

#include <string.h>

#include <gtest/gtest.h>
#include <cstdio>

// TODO: Adapt tests to new changes and remove debug print statements
TEST(TestAx25SendRecv, iFrameLittleStuff) {
  uint8_t telemData[RS_ENCODED_SIZE] = {0};

  unstuffed_ax25_i_frame_t unstuffedAx25Data = {0};
  setCurrentLinkDestAddress(&groundStationCallsign);
  ASSERT_EQ(ax25SendIFrame(telemData, RS_ENCODED_SIZE, &unstuffedAx25Data), OBC_GS_ERR_CODE_SUCCESS);

  packed_ax25_i_frame_t ax25Data = {0};
  ASSERT_EQ(ax25Stuff(unstuffedAx25Data.data, unstuffedAx25Data.length, ax25Data.data, &ax25Data.length),
            OBC_GS_ERR_CODE_SUCCESS);

  unstuffed_ax25_i_frame_t unstuffedPacket = {0};
  ASSERT_EQ(ax25Unstuff(ax25Data.data, ax25Data.length, unstuffedPacket.data, &unstuffedPacket.length),
            OBC_GS_ERR_CODE_SUCCESS);
  EXPECT_EQ(unstuffedPacket.length, AX25_MINIMUM_I_FRAME_LEN);

  u_frame_cmd_t command;
  ASSERT_EQ(ax25Recv(&unstuffedPacket, &command), OBC_GS_ERR_CODE_SUCCESS);
  for (uint16_t i = 0; i < RS_ENCODED_SIZE; ++i) {
    EXPECT_EQ(telemData[i], unstuffedPacket.data[AX25_INFO_FIELD_POSITION + i]);
  }
}

TEST(TestAx25SendRecv, iFrameMaxStuff) {
  uint8_t telemData[RS_ENCODED_SIZE] = {0};
  memset(telemData, 0xFF, RS_ENCODED_SIZE);

  unstuffed_ax25_i_frame_t unstuffedAx25Data = {0};
  setCurrentLinkDestAddress(&groundStationCallsign);
  ASSERT_EQ(ax25SendIFrame(telemData, RS_ENCODED_SIZE, &unstuffedAx25Data), OBC_GS_ERR_CODE_SUCCESS);

  packed_ax25_i_frame_t ax25Data = {0};
  ASSERT_EQ(ax25Stuff(unstuffedAx25Data.data, unstuffedAx25Data.length, ax25Data.data, &ax25Data.length),
            OBC_GS_ERR_CODE_SUCCESS);
  EXPECT_TRUE(ax25Data.length > AX25_MINIMUM_I_FRAME_LEN && ax25Data.length < AX25_MAXIMUM_PKT_LEN);

  unstuffed_ax25_i_frame_t unstuffedPacket = {0};
  ASSERT_EQ(ax25Unstuff(ax25Data.data, ax25Data.length, unstuffedPacket.data, &unstuffedPacket.length),
            OBC_GS_ERR_CODE_SUCCESS);
  EXPECT_EQ(unstuffedPacket.length, AX25_MINIMUM_I_FRAME_LEN);

  u_frame_cmd_t command;
  ASSERT_EQ(ax25Recv(&unstuffedPacket, &command), OBC_GS_ERR_CODE_SUCCESS);
  for (uint16_t i = 0; i < RS_ENCODED_SIZE; ++i) {
    EXPECT_EQ(telemData[i], unstuffedPacket.data[AX25_INFO_FIELD_POSITION + i]);
  }
}

TEST(TestAx25SendRecv, iFrameSomeStuff) {
  uint8_t telemData[RS_ENCODED_SIZE] = {0};
  uint32_t seed = 0xF72FA1;  // random number
  for (uint16_t i = 0; i < RS_ENCODED_SIZE; ++i) {
    // Pseudorandom generation using a simple algorithm
    seed = (seed * 1103515245 + 12345) % (1 << 31);
    telemData[i] = (uint8_t)(seed & 0xFF);
  }

  unstuffed_ax25_i_frame_t unstuffedAx25Data = {0};
  setCurrentLinkDestAddress(&groundStationCallsign);
  ASSERT_EQ(ax25SendIFrame(telemData, RS_ENCODED_SIZE, &unstuffedAx25Data), OBC_GS_ERR_CODE_SUCCESS);

  packed_ax25_i_frame_t ax25Data = {0};
  ASSERT_EQ(ax25Stuff(unstuffedAx25Data.data, unstuffedAx25Data.length, ax25Data.data, &ax25Data.length),
            OBC_GS_ERR_CODE_SUCCESS);

  EXPECT_TRUE(ax25Data.length > AX25_MINIMUM_I_FRAME_LEN && ax25Data.length < AX25_MAXIMUM_PKT_LEN);

  unstuffed_ax25_i_frame_t unstuffedPacket = {0};
  ASSERT_EQ(ax25Unstuff(ax25Data.data, ax25Data.length, unstuffedPacket.data, &unstuffedPacket.length),
            OBC_GS_ERR_CODE_SUCCESS);
  EXPECT_EQ(unstuffedPacket.length, AX25_MINIMUM_I_FRAME_LEN);

  u_frame_cmd_t command;
  ASSERT_EQ(ax25Recv(&unstuffedPacket, &command), OBC_GS_ERR_CODE_SUCCESS);
  for (uint16_t i = 0; i < RS_ENCODED_SIZE; ++i) {
    EXPECT_EQ(telemData[i], unstuffedPacket.data[AX25_INFO_FIELD_POSITION + i]);
  }
}

TEST(TestAx25SendRecv, iFrameSendRecvFlagShare) {
  uint8_t ax25Data[(3 * AX25_MINIMUM_I_FRAME_LEN_SHARE_FLAG) + 1] = {0};
  uint8_t telemData[3 * AX25_INFO_BYTES] = {0};
  uint16_t telemDataLen = (3 * AX25_INFO_BYTES);
  uint16_t ax25DataLen = 0;

  ASSERT_EQ(ax25SendIFrameWithFlagSharing(telemData, telemDataLen, ax25Data,
                                          ax25UnstuffedWithFlagShareLen(telemDataLen), &groundStationCallsign),
            OBC_GS_ERR_CODE_SUCCESS);
  uint8_t stuffedAx25Data[(3 * AX25_MAXIMUM_PKT_LEN)] = {0};
  uint16_t count[3] = {0};

  for (int i = 0; i < 3; ++i) {
    EXPECT_EQ(ax25Data[(AX25_MINIMUM_I_FRAME_LEN_SHARE_FLAG * i)], AX25_FLAG);
    ASSERT_EQ(ax25Stuff(ax25Data + (AX25_MINIMUM_I_FRAME_LEN_SHARE_FLAG * i), AX25_MINIMUM_I_FRAME_LEN,
                        stuffedAx25Data + (AX25_MINIMUM_I_FRAME_LEN_SHARE_FLAG * i), count + i),
              OBC_GS_ERR_CODE_SUCCESS);
  }

  int index = 0;

  uint8_t unstuffedPacket[(3 * AX25_MAXIMUM_PKT_LEN)] = {0};
  uint16_t unstuffedCount[3] = {0};
  int flagCount = 0;
  for (int i = 0; i < 3; ++i) {
    for (int index = 0; index < (3 * AX25_MAXIMUM_PKT_LEN); ++index) {
      if (stuffedAx25Data[i] == AX25_FLAG) {
        ASSERT_EQ(ax25Unstuff(stuffedAx25Data + index, count[i],
                              unstuffedPacket + (AX25_MINIMUM_I_FRAME_LEN_SHARE_FLAG * i), unstuffedCount + i),
                  OBC_GS_ERR_CODE_SUCCESS);
        EXPECT_EQ(unstuffedCount[i], AX25_MINIMUM_I_FRAME_LEN);

        unstuffed_ax25_i_frame_t recvPacket = {0};
        memcpy(recvPacket.data, unstuffedPacket + (AX25_MINIMUM_I_FRAME_LEN_SHARE_FLAG * i), AX25_MINIMUM_I_FRAME_LEN);
        recvPacket.length = AX25_MINIMUM_I_FRAME_LEN;
        u_frame_cmd_t command;
        EXPECT_EQ(ax25Recv(&recvPacket, &command), OBC_GS_ERR_CODE_SUCCESS);
        flagCount++;
      }
      if (flagCount == 4) break;
    }
  }
}

TEST(TestAx25SendRecv, iFrameSendRecvFlagShareStuff) {
  uint8_t ax25Data[(3 * AX25_MINIMUM_I_FRAME_LEN_SHARE_FLAG) + 1] = {0};
  uint8_t telemData[3 * AX25_INFO_BYTES] = {0};
  memset(telemData, 0xFF, 3 * AX25_INFO_BYTES);
  uint16_t telemDataLen = (3 * AX25_INFO_BYTES);
  uint16_t ax25DataLen = 0;

  ASSERT_EQ(ax25SendIFrameWithFlagSharing(telemData, telemDataLen, ax25Data,
                                          ax25UnstuffedWithFlagShareLen(telemDataLen), &groundStationCallsign),
            OBC_GS_ERR_CODE_SUCCESS);
  uint8_t stuffedAx25Data[(3 * AX25_MAXIMUM_PKT_LEN)] = {0};
  uint16_t count[3] = {0};

  for (int i = 0; i < 3; ++i) {
    EXPECT_EQ(ax25Data[(AX25_MINIMUM_I_FRAME_LEN_SHARE_FLAG * i)], AX25_FLAG);
    ASSERT_EQ(ax25Stuff(ax25Data + (AX25_MINIMUM_I_FRAME_LEN_SHARE_FLAG * i), AX25_MINIMUM_I_FRAME_LEN,
                        stuffedAx25Data + (AX25_MINIMUM_I_FRAME_LEN_SHARE_FLAG * i), count + i),
              OBC_GS_ERR_CODE_SUCCESS);
  }

  int index = 0;

  uint8_t unstuffedPacket[(3 * AX25_MAXIMUM_PKT_LEN)] = {0};
  uint16_t unstuffedCount[3] = {0};
  int flagCount = 0;
  for (int i = 0; i < 3; ++i) {
    for (int index = 0; index < (3 * AX25_MAXIMUM_PKT_LEN); ++index) {
      if (stuffedAx25Data[i] == AX25_FLAG) {
        ASSERT_EQ(ax25Unstuff(stuffedAx25Data + index, count[i],
                              unstuffedPacket + (AX25_MINIMUM_I_FRAME_LEN_SHARE_FLAG * i), unstuffedCount + i),
                  OBC_GS_ERR_CODE_SUCCESS);
        EXPECT_EQ(unstuffedCount[i], AX25_MINIMUM_I_FRAME_LEN);

        unstuffed_ax25_i_frame_t recvPacket = {0};
        memcpy(recvPacket.data, unstuffedPacket + (AX25_MINIMUM_I_FRAME_LEN_SHARE_FLAG * i), AX25_MINIMUM_I_FRAME_LEN);
        recvPacket.length = AX25_MINIMUM_I_FRAME_LEN;
        u_frame_cmd_t command;
        EXPECT_EQ(ax25Recv(&recvPacket, &command), OBC_GS_ERR_CODE_SUCCESS);
        flagCount++;
      }
      if (flagCount == 4) break;
    }
  }
}

TEST(TestAx25SendRecv, uFrameSendRecvConn) {
  packed_ax25_u_frame_t ax25Data = {0};
  uint8_t pollFinalBit = 1;

  setCurrentLinkDestAddress(&cubesatCallsign);
  ASSERT_EQ(ax25SendUFrame(&ax25Data, U_FRAME_CMD_CONN, pollFinalBit), OBC_GS_ERR_CODE_SUCCESS);

  for (int i = 0; i < ax25Data.length; i++) {
    printf(" 0x%x", ax25Data.data[i]);
  }

  unstuffed_ax25_i_frame_t unstuffedPacket = {0};
  ASSERT_EQ(ax25Unstuff(ax25Data.data, ax25Data.length, unstuffedPacket.data, &unstuffedPacket.length),
            OBC_GS_ERR_CODE_SUCCESS);

  u_frame_cmd_t command;
  ASSERT_EQ(ax25Recv(&unstuffedPacket, &command), OBC_GS_ERR_CODE_SUCCESS);
  EXPECT_EQ(command, U_FRAME_CMD_CONN);
}

TEST(TestAx25SendRecv, uFrameSendRecvDisc) {
  packed_ax25_u_frame_t ax25Data = {0};
  uint8_t pollFinalBit = 1;

  setCurrentLinkDestAddress(&cubesatCallsign);
  ASSERT_EQ(ax25SendUFrame(&ax25Data, U_FRAME_CMD_DISC, pollFinalBit), OBC_GS_ERR_CODE_SUCCESS);

  unstuffed_ax25_i_frame_t unstuffedPacket = {0};
  ASSERT_EQ(ax25Unstuff(ax25Data.data, ax25Data.length, unstuffedPacket.data, &unstuffedPacket.length),
            OBC_GS_ERR_CODE_SUCCESS);

  u_frame_cmd_t command;
  ASSERT_EQ(ax25Recv(&unstuffedPacket, &command), OBC_GS_ERR_CODE_SUCCESS);
  EXPECT_EQ(command, U_FRAME_CMD_DISC);
}

TEST(TestAx25SendRecv, uFrameSendRecvAck) {
  packed_ax25_u_frame_t ax25Data = {0};
  uint8_t pollFinalBit = 1;

  setCurrentLinkDestAddress(&cubesatCallsign);
  ASSERT_EQ(ax25SendUFrame(&ax25Data, U_FRAME_CMD_ACK, pollFinalBit), OBC_GS_ERR_CODE_SUCCESS);

  unstuffed_ax25_i_frame_t unstuffedPacket = {0};
  ASSERT_EQ(ax25Unstuff(ax25Data.data, ax25Data.length, unstuffedPacket.data, &unstuffedPacket.length),
            OBC_GS_ERR_CODE_SUCCESS);

  u_frame_cmd_t command;
  ASSERT_EQ(ax25Recv(&unstuffedPacket, &command), OBC_GS_ERR_CODE_SUCCESS);
  EXPECT_EQ(command, U_FRAME_CMD_ACK);
}

TEST(TestAx25SendRecV, Ax25SourceAddressGenerator) {
  ax25_addr_t sourceAddress;
  memset(sourceAddress.data, 0, 7);
  sourceAddress.length = 7;

  uint8_t expectedAddress[] = {0x9C, 0x6E, 0x98, 0x8A, 0x9A, 0x40, 0x61};  // Source subfield from documentation
  uint8_t callSign[] = {'N', '7', 'L', 'E', 'M'};
  uint8_t callSign2[] = {'N', '7', 'L', 'E', 'M', '4', '2'};

  ASSERT_EQ(ax25GetSourceAddress(&sourceAddress, callSign2, 7, 0, 0), 1);

  ax25GetSourceAddress(&sourceAddress, callSign, 5, 0, 0);
  ASSERT_EQ(memcmp(&sourceAddress, &expectedAddress, 7), 0);
}

TEST(TestAx25SendRecV, Ax25DestAddressGenerator) {
  ax25_addr_t sourceAddress;
  memset(sourceAddress.data, 0, 7);
  sourceAddress.length = 7;

  uint8_t expectedAddress[] = {0x9C, 0x94, 0x6E, 0xA0, 0x40, 0x40, 0xE0};  // Source subfield from documentation
  uint8_t callSign[] = {'N', 'J', '7', 'P'};
  uint8_t callSign2[] = {'N', '7', 'L', 'E', 'M', '4', '2'};

  ASSERT_EQ(ax25GetDestAddress(&sourceAddress, callSign2, 7, 0, 0), 1);
  ax25GetDestAddress(&sourceAddress, callSign, 4, 0, 1);
  ASSERT_EQ(memcmp(&sourceAddress, &expectedAddress, 7), 0);
}

File: python_test/interface_wrapper_tests/test_obc_gs_ax25.py
import interfaces.obc_gs_interface.python_wrappers.error_codes.obc_gs_errors as errors
import interfaces.obc_gs_interface.python_wrappers.rs_encode_decode.obc_gs_fec as fec
import interfaces.obc_gs_interface.python_wrappers.uplink_flow_encode.obc_gs_ax25 as ax25
import pytest


@pytest.fixture
def telem_data():
    """Returns a valid packed telemetry data bytearray."""
    return bytearray(fec.RS_ENCODED_SIZE)


@pytest.fixture
def unstuffed_ax25_data():
    """Returns a valid unstuffed AX25 I-frame as a dictionary."""
    return {"data": bytearray(ax25.AX25_MINIMUM_I_FRAME_LEN), "length": 0}


@pytest.fixture
def packed_ax25_u_frame():
    """Returns a valid packed AX25 U-frame as a dictionary."""
    return {"data": bytearray(ax25.AX25_MAXIMUM_U_FRAME_CMD_LENGTH), "length": 0}


@pytest.fixture
def cubsat_call_sign():
    """Returns a valid packed AX25 U-frame as a dictionary."""
    return {"data": bytearray(ax25.AX25_DEST_ADDR_BYTES), "length": ax25.AX25_DEST_ADDR_BYTES}


@pytest.fixture
def groundstation_call_sign():
    """Returns a valid packed AX25 U-frame as a dictionary."""
    return {"data": bytearray(ax25.AX25_DEST_ADDR_BYTES), "length": ax25.AX25_DEST_ADDR_BYTES}


def test_i_frame_little_stuff(telem_data, unstuffed_ax25_data, groundstation_call_sign):
    """Test sending and receiving an I-frame with little bit stuffing."""
    ax25.setCurrentLinkDestAddress(groundstation_call_sign)
    result = ax25.ax25SendIFrame(telem_data, len(telem_data), unstuffed_ax25_data)
    assert result == errors.OBC_GS_ERR_CODE_SUCCESS

    packed_ax25_data = {"data": bytearray(ax25.AX25_MAXIMUM_PKT_LEN), "length": 0}
    result = ax25.ax25Stuff(
        unstuffed_ax25_data["data"], unstuffed_ax25_data["length"], packed_ax25_data["data"], packed_ax25_data["length"]
    )
    assert result == errors.OBC_GS_ERR_CODE_SUCCESS

    unstuffed_packet = {"data": bytearray(ax25.AX25_MINIMUM_I_FRAME_LEN), "length": 0}
    result = ax25.ax25Unstuff(
        packed_ax25_data["data"], packed_ax25_data["length"], unstuffed_packet["data"], unstuffed_packet["length"]
    )
    assert result == errors.OBC_GS_ERR_CODE_SUCCESS
    assert unstuffed_packet["length"] == ax25.AX25_MINIMUM_I_FRAME_LEN

    command = ax25.U_FRAME_CMD_CONN
    result = ax25.ax25Recv(unstuffed_packet, command)
    assert result == errors.OBC_GS_ERR_CODE_SUCCESS

    for i in range(fec.RS_ENCODED_SIZE):
        assert telem_data[i] == unstuffed_packet["data"][ax25.AX25_INFO_FIELD_POSITION + i]


# def test_i_frame_max_stuff(telem_data, unstuffed_ax25_data, groundstation_call_sign):
#     """Test sending and receiving an I-frame with maximum bit stuffing."""
#     telem_data[:] = [0xFF] * fec.RS_ENCODED_SIZE
#
#     ax25.setCurrentLinkDestAddress(groundstation_call_sign)
#     result = ax25.ax25SendIFrame(telem_data, len(telem_data), unstuffed_ax25_data)
#     assert result == errors.OBC_GS_ERR_CODE_SUCCESS
#
#     packed_ax25_data = {"data": bytearray(ax25.AX25_MAXIMUM_PKT_LEN), "length": 0}
#     result = ax25.ax25Stuff(unstuffed_ax25_data["data"], unstuffed_ax25_data["length"],
#                             packed_ax25_data["data"], packed_ax25_data["length"])
#     assert result == errors.OBC_GS_ERR_CODE_SUCCESS
#     assert ax25.AX25_MINIMUM_I_FRAME_LEN < packed_ax25_data["length"] < ax25.AX25_MAXIMUM_PKT_LEN
#
#     unstuffed_packet = {"data": bytearray(ax25.AX25_MINIMUM_I_FRAME_LEN), "length": 0}
#     result = ax25.ax25Unstuff(packed_ax25_data["data"], packed_ax25_data["length"],
#                               unstuffed_packet["data"], unstuffed_packet["length"])
#     assert result == errors.OBC_GS_ERR_CODE_SUCCESS
#     assert unstuffed_packet["length"] == ax25.AX25_MINIMUM_I_FRAME_LEN
#
#     command = ax25.ax25Recv(unstuffed_packet)
#     assert command == errors.OBC_GS_ERR_CODE_SUCCESS
#
#     for i in range(fec.RS_ENCODED_SIZE):
#         assert telem_data[i] == unstuffed_packet["data"][ax25.AX25_INFO_FIELD_POSITION + i]
#
# def test_i_frame_some_stuff(telem_data, unstuffed_ax25_data, groundstation_call_sign):
#     """Test sending and receiving an I-frame with some bit stuffing."""
#     seed = 0xF72FA1
#     for i in range(fec.RS_ENCODED_SIZE):
#         seed = (seed * 1103515245 + 12345) & 0x7FFFFFFF
#         telem_data[i] = seed & 0xFF
#
#     ax25.setCurrentLinkDestAddress(groundstation_call_sign)
#     result = ax25.ax25SendIFrame(telem_data, len(telem_data), unstuffed_ax25_data)
#     assert result == errors.OBC_GS_ERR_CODE_SUCCESS
#
#     packed_ax25_data = {"data": bytearray(ax25.AX25_MAXIMUM_PKT_LEN), "length": 0}
#     result = ax25.ax25Stuff(unstuffed_ax25_data["data"], unstuffed_ax25_data["length"],
#                             packed_ax25_data["data"], packed_ax25_data["length"])
#     assert result == errors.OBC_GS_ERR_CODE_SUCCESS
#     assert ax25.AX25_MINIMUM_I_FRAME_LEN < packed_ax25_data["length"] < ax25.AX25_MAXIMUM_PKT_LEN
#
#     unstuffed_packet = {"data": bytearray(ax25.AX25_MINIMUM_I_FRAME_LEN), "length": 0}
#     result = ax25.ax25Unstuff(packed_ax25_data["data"], packed_ax25_data["length"],
#                               unstuffed_packet["data"], unstuffed_packet["length"])
#     assert result == errors.OBC_GS_ERR_CODE_SUCCESS
#     assert unstuffed_packet["length"] == ax25.AX25_MINIMUM_I_FRAME_LEN
#
#     command = ax25.ax25Recv(unstuffed_packet)
#     assert command == errors.OBC_GS_ERR_CODE_SUCCESS
#
#     for i in range(fec.RS_ENCODED_SIZE):
#         assert telem_data[i] == unstuffed_packet["data"][ax25.AX25_INFO_FIELD_POSITION + i]
#
# def test_u_frame_send_recv_conn(packed_ax25_u_frame, cubsat_call_sign):
#     """Test sending and receiving a U-frame with CONN command."""
#     poll_final_bit = 1
#
#     ax25.setCurrentLinkDestAddress(cubsat_call_sign)
#     result = ax25.ax25SendUFrame(packed_ax25_u_frame, ax25.U_FRAME_CMD_CONN, poll_final_bit)
#     assert result == errors.OBC_GS_ERR_CODE_SUCCESS
#
#     unstuffed_packet = {"data": bytearray(ax25.AX25_MINIMUM_I_FRAME_LEN), "length": 0}
#     result = ax25.ax25Unstuff(packed_ax25_u_frame["data"], packed_ax25_u_frame["length"],
#                               unstuffed_packet["data"], unstuffed_packet["length"])
#     assert result == errors.OBC_GS_ERR_CODE_SUCCESS
#
#     command = ax25.ax25Recv(unstuffed_packet)
#     assert command == ax25.U_FRAME_CMD_CONN
#
# def test_u_frame_send_recv_disc(packed_ax25_u_frame, cubsat_call_sign):
#     """Test sending and receiving a U-frame with DISC command."""
#     poll_final_bit = 1
#
#     ax25.setCurrentLinkDestAddress(cubsat_call_sign)
#     result = ax25.ax25SendUFrame(packed_ax25_u_frame, ax25.U_FRAME_CMD_DISC, poll_final_bit)
#     assert result == errors.OBC_GS_ERR_CODE_SUCCESS
#
#     unstuffed_packet = {"data": bytearray(ax25.AX25_MINIMUM_I_FRAME_LEN), "length": 0}
#     result = ax25.ax25Unstuff(packed_ax25_u_frame["data"], packed_ax25_u_frame["length"],
#                               unstuffed_packet["data"], unstuffed_packet["length"])
#     assert result == errors.OBC_GS_ERR_CODE_SUCCESS
#
#     command = ax25.ax25Recv(unstuffed_packet)
#     assert command == ax25.U_FRAME_CMD_DISC
#
# def test_ax25_source_address_generator():
#     """Test generating AX25 source address."""
#     source_address = {"data": bytearray(7), "length": 7}
#     expected_address = bytearray([0x9C, 0x6E, 0x98, 0x8A, 0x9A, 0x40, 0x61])
#     call_sign = b'N7LEM'
#     call_sign2 = b'N7LEM42'
#
#     result = ax25.ax25GetSourceAddress(source_address, call_sign2, 7, 0, 0)
#     assert result == 1
#
#     ax25.ax25GetSourceAddress(source_address, call_sign, 5, 0, 0)
#     assert source_address["data"] == expected_address
#
# def test_ax25_dest_address_generator():
#     """Test generating AX25 destination address."""
#     dest_address = {"data": bytearray(7), "length": 7}
#     expected_address = bytearray([0x9C, 0x94, 0x6E, 0xA0, 0x40, 0x40, 0xE0])
#     call_sign = b'NJ7P'
#     call_sign2 = b'N7LEM42'
#
#     result = ax25.ax25GetDestAddress(dest_address, call_sign2, 7, 0, 0)
#     assert result == 1
#
#     ax25.ax25GetDestAddress(dest_address, call_sign, 4, 0, 1)
#     assert dest_address["data"] == expected_address
